File: python-workspace/백준/다이나믹프로그래밍/RGB거리.py
import sys

count_of_house = int(sys.stdin.readline())
cost = []
for _ in range(count_of_house):
    cost.append(list(map(int, sys.stdin.readline().split(" "))))
dp = [[float('inf') for _ in range(3)] for _ in range(count_of_house)]
dp[0][0] = cost[0][0]
dp[0][1] = cost[0][1]
dp[0][2] = cost[0][2]

for house in range(1, count_of_house):
    dp[house][0] = cost[house][0] + min(dp[house - 1][1], dp[house - 1][2])
    dp[house][1] = cost[house][1] + min(dp[house - 1][0], dp[house - 1][2])
    dp[house][2] = cost[house][2] + min(dp[house - 1][0], dp[house - 1][1])

answer = min(dp[count_of_house - 1][0], dp[count_of_house - 1][1], dp[count_of_house - 1][2])
print(answer)

File: java-workspace3/src/다이나믹프로그래밍/RGB거리.java
package 다이나믹프로그래밍;

import java.util.Scanner;

public class RGB거리 {
	static Scanner scanner = new Scanner(System.in);
	static int[][] costs;

	public static void main(String[] args) {
		int countOfHouse = scanner.nextInt();
		costs = new int[countOfHouse][4];
		for (int house = 0; house < countOfHouse; house++) {
			for (int color = 1; color <= 3; color++) {
				costs[house][color] = scanner.nextInt();
			}
		}
		int answer = solution(countOfHouse);
		System.out.println(answer);
	}

	private static int solution(int countOfHouse) {
		int[][] dp = new int[countOfHouse][5];
		int INF = Integer.MAX_VALUE;
		int answer = INF;
		for (int house = 0; house < countOfHouse; house++) {
			for (int color = 1; color <= 3; color++) {
				dp[house][color] = INF;
			}
		}
		for (int color = 1; color <= 3; color++) {
			dp[0][color] = costs[0][color];
		}
		for (int house = 1; house < countOfHouse; house++) {
			for (int pivot = 1; pivot <= 3; pivot++) {
				for (int compare = 1; compare <= 3; compare++) {
					if (pivot == compare) {
						continue;
					}
					dp[house][pivot] = Math.min(dp[house][pivot], dp[house - 1][compare] + costs[house][pivot]);
				}
			}
		}
		for (int color = 1; color <= 3; color++) {
			answer = Math.min(dp[countOfHouse - 1][color], answer);
		}
		return answer;
	}
}
